astropy/io/ascii/core.py

```diff
@@ -1345,8 +1345,17 @@ def _get_reader(Reader, Inputter=None, Outputter=None, **kwargs):
             kwargs['Inputter'] = Inputter
         return Reader(**kwargs)
 
+    # If user explicitly passed a fast reader with 'force' or with non-default
+    # options for the fast reader, raise an error for slow readers
     if 'fast_reader' in kwargs:
-        del kwargs['fast_reader'] # ignore fast_reader parameter for slow readers
+        if kwargs['fast_reader'] == 'force' or \
+           isinstance(kwargs['fast_reader'] , dict):
+            raise ParameterError('fast_reader required with ' +
+                                 '{0}, but this is a slow reader: {1}'
```

Comment:
Well, I had left this on the shelf far too long myself...
And the discussion became scattered a bit across different PRs and issues like [5574](https://github.com/astropy/astropy/issues/5574#issuecomment-265896576) - basically there some tests in `test_read` were arguably wrong, as they presumed to test the `fast_reader` and expected an `InconsistentTableError`, which was not implemented in the C parser. So one option I suggested there was to conditionally check for either `InconsistentTableError` or `CParserError` depending on the `fast_reader` setting, or implement `InconsistentTableError` for this case. It seemed preferable to have consistent errors returned as far as possible, regardless of the underlying engine used, but of course this is open to discussion.